src/statsd-parsers/basic/basic.c

```diff
@@ -194,31 +205,45 @@ int parse(char* buffer, statsd_datagram** datagram) {
     error_clean_up:
     free(attr);
     free(segment);
-
-    error_clean_up_end:
-    if (field_allocated_flags & 1) {
-        free_datagram_tags((*datagram)->tags);
-    }
-    if (field_allocated_flags & 1 << 1) {
-        free((*datagram)->metric);
-    }
-    if (field_allocated_flags & 1 << 2) {
-        free((*datagram)->type);
-    }
-    if (field_allocated_flags & 1 << 3) {
-        free((*datagram)->value);
-    }
-    if (field_allocated_flags & 1 << 4) {
-        free((*datagram)->instance);
-    }
-    if (field_allocated_flags & 1 << 5) {
-        free((*datagram)->sampling);
-    }
     if (tag_allocated_flags & 1) {
         free(tag_key);
     }
     if (tag_allocated_flags & 2) {
         free(tag_value);
     }
+    error_clean_up_end:
     return 0;
+}
+
+static int tag_comparator(const void* x, const void* y) {
+    int res = strcmp((*(tag**)x)->key, (*(tag**)y)->key);
+    return res;
+}
+
+/**
+ * Converts tag_collection* struct to JSON string that is sorted by keys
+ */
+char* tag_collection_to_json(tag_collection* tags) {
+    char buffer[JSON_BUFFER_SIZE];
+    memset(buffer, '\0', JSON_BUFFER_SIZE);
+    qsort(tags->values, tags->length, sizeof(tag*), tag_comparator);
+    buffer[0] = '{';
+    int i;
+    int current_size;
+    for (i = 0; i < tags->length; i++) {
+        current_size = strlen(buffer);
+        if (i == 0) {
+            snprintf(buffer + current_size, JSON_BUFFER_SIZE - current_size, "\"%s\":\"%s\"",
+                tags->values[i]->key, tags->values[i]->value);
+        } else {
+            snprintf(buffer + current_size, JSON_BUFFER_SIZE - current_size, ",\"%s\":\"%s\"",
+                tags->values[i]->key, tags->values[i]->value);
+        }
+    }
+    current_size = strlen(buffer);
+    buffer[current_size] = '}';
```

Comment:
One extra check before the final } (and null char) are appended is needed too.

Reply:
Fixed in b92b51c